src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,271 @@
+        if let Some(def) = mplace.layout.ty.ty_adt_def() {
+            if Some(def.did) == self.ecx.tcx.lang_items().unsafe_cell_type() {
+                // we are crossing over an `UnsafeCell`, const qualif ensures that this can only
+                // exist in the base constant, not behind references. Double checking here to make
+                // sure everything is alright
+                assert_eq!(self.mutability, Mutability::Mutable);
+            }
+        }
+        self.walk_aggregate(mplace, fields)
+    }
+
+    fn visit_primitive(&mut self, mplace: MPlaceTy<'tcx>) -> EvalResult<'tcx> {
+        // Handle Reference types, as these are the only relocations supported by const eval.
+        // Raw pointers (and boxes) are handled by the `leftover_relocations` logic.
+        let ty = mplace.layout.ty;
+        if let ty::Ref(_, _, mutability) = ty.sty {
+            let value = self.ecx.read_immediate(mplace.into())?;
+            // Handle trait object vtables
+            if let Ok(meta) = value.to_meta() {
+                let layout = self.ecx.layout_of(ty.builtin_deref(true).unwrap().ty)?;
+                if layout.is_unsized() {
+                    if let ty::Dynamic(..) = self.ecx.tcx.struct_tail(layout.ty).sty {
+                        if let Ok(vtable) = meta.unwrap().to_ptr() {
+                            // explitly choose `Immutable` here, since vtables are immutable, even
+                            // if the reference of the fat pointer is mutable
+                            self.intern(vtable, Mutability::Immutable)?;
+                        }
+                    }
+                }
+            }
+            let mplace = self.ecx.ref_to_mplace(value)?;
+            // Check if we have encountered this pointer+layout combination before.
+            // Only recurse for allocation-backed pointers.
+            if let Scalar::Ptr(ptr) = mplace.ptr {
+                // In the future we will probably allow `& &mut T`, and thus will want to merge
+                // `mutability` with `self.mutability` to only choose `Mutable` if both are
+                // `Mutable`.
+
+                // We do not have any `frozen` logic here, because it's essentially equivalent to
+                // the mutability except for the outermost item. Only `UnsafeCell` can "unfreeze",
+                // and we check that in `visit_aggregate`.
+                match (self.mode, mutability) {
+                    // all is "good and well" in the unsoundness of `static mut`
+                    (InternMode::StaticMut, _) => {},
+                    // immutable references are fine everywhere
+                    (InternMode::StaticOrConst, hir::Mutability::MutImmutable) => {},
+                    // we statically prevent `&mut T` via `const_qualif`
+                    (InternMode::StaticOrConst, hir::Mutability::MutMutable) =>
+                        bug!("const qualif failed to prevent mutable references"),
```

Comment:
You haven't replied to this when I brought it up elsewhere so I'll ask again. :)  I find it somewhat strange that we are forbidding `&mut` in `static` here. There's nothing unsound about that, and if we did anything wrong elsewhere we'd ICE. This is in strong contrast to `&mut` in `const`, we surely want nothing mutable in there.  By conflating these two cases (one limitation is fundamental and one is just an artifact of earlier passes in the compiler), I find this harder to follow.

Reply:
Hmm.. you mean because accessing a `static` is already going through a `&` at the access site essentially?

Reply:
Yes.

Reply:
Added a test and made it not ICE (test requires -Zunleash-the-miri-inside-of-you)